src/bank.rs

```diff
@@ -246,149 +252,238 @@ impl Bank {
     /// Deduct tokens from the 'from' address the account has sufficient
     /// funds and isn't a duplicate.
     fn apply_debits(
-        &self,
         tx: &Transaction,
-        accounts: &mut HashMap<Pubkey, Account>,
-        vote_no_account_err: &mut usize,
-        no_account_err: &mut usize,
+        accounts: &mut [Account],
+        instruction: &Instruction,
     ) -> Result<()> {
-        let mut purge = false;
         {
-            let option = accounts.get_mut(&tx.from);
-            if option.is_none() {
-                if let Instruction::NewVote(_) = &tx.instruction {
-                    *vote_no_account_err += 1;
-                } else {
-                    *no_account_err += 1;
-                }
-                return Err(BankError::AccountNotFound(tx.from));
-            }
-            let bal = option.unwrap();
-
-            self.reserve_signature_with_last_id(&tx.signature, &tx.last_id)?;
-
-            if let Instruction::NewContract(contract) = &tx.instruction {
+            let empty = accounts[0].userdata.is_empty();
+            let tokens = if !empty { 0 } else { accounts[0].tokens };
```

Comment:
why do we ignore account[0].tokens when there's no user data?

Reply:
@mvines you are right, that doesn't make any sense.